GoMeet/src/main/webapp/js/meeting-location-dao.js

```diff
@@ -1,26 +1,128 @@
+/** Constructs a Data Access Object for location data. */
+class MeetingLocationDaoFactory {
+  /** 
+   * Returns a meetingLocationDAO object based on whether the map stores to a temporary
+   * or permanent location.
+   */
+  static getLocationDao(storageType) {
+    if (storageType === 'permanent') {
+      return new PermMeetingLocationDao();
+    } else {
+      return new TempMeetingLocationDao();
+    }
+  }
+}
+
+/** Stores location data to session storage. */
+class TempMeetingLocationDao {
+  constructor() {
+    this.emptyKeyString = '';
+    this.MAX_NUM_ENTITIES = 5;
+  }
+
+  /** Fetches the location data stored in session storage. */
+  fetchLocations() {
+    if (sessionStorage.getItem('locations') == null) {
+      return [];
+    } else {
+      return JSON.parse(sessionStorage.getItem('locations'));
+    } 
+  }
+
+  /** Saves a new location to session storage. */
+  newLocation(title, lat, lng, note) {
+    // Check if the new location can be added.
+    if (title === '') {
+      throw new Error(BLANK_FIELDS_ALERT);
+    }
+    if (!this.validTitle(title)) {
+      throw new Error(SAME_TITLE);
+    }
+    if (this.maxEntitiesReached()) {
+      throw new Error(MAX_ENTITIES);
+    }
+
+    // Add the new location.
+    let locationList;
+    if (sessionStorage.getItem('locations') == null) {
+      locationList = [];
+    } else {
+      locationList = JSON.parse(sessionStorage.getItem('locations'));
+    }
+    const newLocation = {title: title, lat: lat, lng: lng,
+        note: note, keyString: this.emptyKeyString};
+    locationList.push(newLocation);
+    sessionStorage.setItem('locations', JSON.stringify(locationList));
+  }
+
+  /**
+   * Called when the user clicks the VOTE button.
+   * Note: Temporary storage is only used during the meeting creation.
+   * The meeting creator cannot vote more than once.
+   * @throws an error when the meeting creator tries to vote for location.
+   */
+  updateLocation(keyString) {
+    throw new Error(USER_HAS_VOTED);
+  }
+
+  /**
+   * Checks if the title is unique and therefore valid. i.e.
+   * There is no other location for this meeting with the same title.
+   * @param {String} targetTitle the title of the location being added.
+   * @return true if the title is valid.
+   */
+  validTitle(targetTitle) {
+    const locations = this.fetchLocations();
+    if (locations.length === 0) {
+      return true;
+    }
+    for (let i = 0; i < locations.length; i++) {
```

Comment:
Optional: Potentially to reduce the complexity of this function, every time a user adds a location, add the location name to a set that persists while the add location page is displayed (e.g. global variable in the script), and then pass that set to this function, check whether the new location name is in the set to determine if it is unique..

Reply:
Yup, I think since the number of locations is capped at 5, it's okay for now. But that way would be more efficient for large datasets.